branchsearch.cpp

```diff
@@ -93,6 +93,40 @@ void filter_branches(vector<string> &all_branches, char* regex_value, vector<str
     }
 }
 
+bool best_match(vector<string> &all_branches, char* pattern, string& res){
+    // pattern to lowercase
+    std::transform(pattern, pattern+strlen(pattern), pattern, [](unsigned char c){ return std::tolower(c); });
+
+    // take the ones containing the pattern
+    vector<string*> matches;
+    for (string &b: all_branches){
+        // branch to lowercase
+        std::transform(b.begin(), b.end(), b.begin(), [](unsigned char c){ return std::tolower(c); });
+        if (b.find(pattern) != string::npos) matches.push_back(&b);
+    }
+
+    if(!matches.size()) return false;
+
+    // sort by some similarity criteria
+    sort(matches.begin(),matches.end(), [pattern]( const string* a, const string* b)
+        {
+            // the one starting with the pattern is prefered
+            if(a->find(pattern) == 0 and b->find(pattern) != 0) return true;
+            if(b->find(pattern) == 0 and a->find(pattern) != 0) return false;
+
+            // the one closer in length to the pattern is preferd
+            if(a->size()-strlen(pattern) < b->size()-strlen(pattern)) return true;
+            if(b->size()-strlen(pattern) < a->size()-strlen(pattern)) return false;
+
+            // tiebreaker
+            return *a < *b;
+        }
+    );
+
+    res = *matches[0]; // only care about the best match here
+    return true;
```

Comment:
From what I read, it is possible not to return something. Doesn't look like it here, but I think the possibility of not matching anything should be valid.

Reply:
Didn’t follow you there. The boolean the function I returning is to indicate whether a branch was found and stored in “res”.

Reply:
Why not just return the string itself? A single copy shouldn't be that bad.

Reply:
What would you return if there is no match? An empty string? 🤢

Reply:
Indeed. Why would that be so bad? It matched that, nothing.

Reply:
>  It matched that, nothing, epsilon.

No it didn’t. This function is looking for a string in the vector that contains the pattern. An empty string does not contain the pattern and is probably not in the vector. 

I think returning an empty string is hacky and sacrifices generality. Probably not a problem in this small project, but if I already avoided it...

Reply:
I struggle to see how the current solution is more general, but ok.